### DISCUSSION TITLE
mpv使用yt-dlp播放B站视频问题解决方法

### DISCUSSION THREAD

**Discussion by zhongfly** [hooke007/mpv_PlayKit#94] [mpv 通用]
目前mpv使用yt-dlp播放B站视频存在2个问题
1.加载视频时不使用yt-dlp提供的http_headers，导致加载视频失败（403错误）
2.把弹幕（xml）当做字幕加载时，播放器卡死，且无选项使yt-dlp排除xml格式

基本上都与ytdl_hook.lua有关，如果解决也需要从此下手
1的话，https://github.com/mpv-player/mpv/issues/9978 ，在脚本579行后添加新一行`set_http_headers(json.http_headers)`，就能解决问题，不过也可以有更好的解决办法
2的话，我尝试修改脚本传给yt-dlp的--sub-format选项，发现问题不在这里，而在于当不存在指定格式字幕时，yt-dlp会自动选择一种字幕，即使它违反了sub-format的要求。https://github.com/yt-dlp/yt-dlp/issues/3058 yt-dlp已列入计划

**Comment by zhongfly** [comment]
我自己的修改版
https://gist.github.com/zhongfly/e95fa433ca912380f9f61e0910146d7e

@dyphire

**Reply by dyphire** [reply]
感觉提个pr比较好

**Reply by zhongfly** [reply]
> 感觉提个pr比较好

1可以，但2的修改可能就不清楚，我不清楚为啥要用`all-subs`而不是`--sub-lang all`

**Reply by dyphire** [reply]
> 1可以，但2的修改可能就不清楚，我不清楚为啥要用`all-subs`而不是`--sub-lang all`

感觉是他们一直没注意到`--all-subs`和`--sub-format`间的覆盖冲突，就算选择`--all-subs`，也应该在它之后添加`--sub-format`

暂时先合并到我的修改版mpv了 https://github.com/dyphire/mpv

**Reply by zhongfly** [reply]
选项的顺序有意义？

**Reply by dyphire** [reply]
没测试，不确定是否顺序覆盖（

但是`--all-subs`和`--sub-format`两者存在冲突是确定的，那么原脚本的这两个参数就不合理了。issue或pr都挺合适

**Reply by zhongfly** [reply]
看了下yt-dlp的文档，我用的`.http_headers`不是文档中记录的，只是有部分网站解析会这么做，正规的方法都是在`.requested_formats[].http_headers`里的，然后多个流文件中就会有对应的多个headers信息，但播放时mpv只能为整个视频设置一个headers，问题就变成脚本应该选择哪一个，而且怎么改代码（复杂不少）

目前简单改着应付bilibili就行了

**Reply by dyphire** [reply]
> 1可以，但2的修改可能就不清楚，我不清楚为啥要用`all-subs`而不是`--sub-lang all`

测试了下，知道为什么使用`--all-subs`了，不使用此选项的话yt-dlp是不会下载字幕的，用youtube测试使用`--sub-lang all`后vtt字幕没有加载，使用`--all-subs`后方可加载

所以2确实不应该改

**Reply by zhongfly** [reply]
> 不使用此选项的话yt-dlp是不会下载字幕的

本来就不是yt-dlp下载字幕，yt-dlp只负责解析

**Reply by zhongfly** [reply]
测试应该使用`yt-dlp --no-warnings -J --flat-playlist --format bestvideo+bestaudio/best --sub-format <字幕格式> --sub-lang all --no-playlist -- <视频url>`，然后使用json格式化工具格式化，其中requested_subtitles是mpv会添加的字幕轨

**Reply by dyphire** [reply]
> 本来就不是yt-dlp下载字幕，yt-dlp只负责解析

我是指mpv使用yt-dlp的字幕加载行为。事实是`--write-subs`影响mpv的字幕加载，而使用`--all-subs`时`--sub-format`选项会冲突。
mpv使用youtube-dl时不加载字幕的原因也是这个，youtube-dl的`--all-subs`选项并不包含`--write-subs`，所以必须单独传递`--write-sub`参数youtube-dl才会为mpv写入字幕

所以要么加入以下修改
```lua
    local command = {
        ytdl.path, "--no-warnings", "-J", "--write-subs", "--flat-playlist"
    }
```
要么设置`--ytdl-raw-options-append=write-subs=`

**Reply by zhongfly** [reply]
应该是yt-dlp有问题

**Reply by dyphire** [reply]
> 应该是yt-dlp有问题

可能是，不过`--write-subs`和`--sub-format`可以一起使用。坑爹的是youtube-dl的参数名称不一样`--write-sub`，不过我用`--write-subs`测试youtube-dl也可以正常加载字幕，迷惑

测了下yt-dlp也识别`--write-sub`选项
所以我觉得使用`--ytdl-raw-options=`选项来指定要比较好

**Reply by zhongfly** [reply]
https://github.com/yt-dlp/yt-dlp/issues/3058 看看yt-dlp怎么说，也许人家就是这么设计的

**Reply by zhongfly** [reply]
看了下代码，他就是这么设计的，没有合适的，就用该语言的最后一个字幕…看来yt-dlp得为此加一个新选项才能解决

**Comment by zhongfly** [comment]
如果不修改脚本的话，1可以用自动配置解决（只针对B站）
```ini
[bilibili-add-referrer]
 profile-desc = Add referrer for BiliBili video
 profile-cond = path:match("https://www.bilibili.com")~=nil
 profile-restore = copy
 referrer = https://www.bilibili.com/
```

**Comment by zhongfly** [comment]
https://github.com/mpv-player/mpv/commit/431473310f9d86f6ae030ce3432897edebe5ab89 应该是被修复了

**Reply by dyphire** [reply]
确实修复了，不过我更关心他的其他几个pr。ytdl的几个暂时被我合并到修改版[ytdl_hook_plus.lua](https://github.com/dyphire/mpv-scripts/blob/main/ytdl_hook_plus.lua)了
